conans/client/tools/win.py

```diff
@@ -63,22 +64,71 @@ def build_sln_command(settings, sln_path, targets=None, upgrade_project=True, bu
     return command
 
 
-def vs_installation_path(version):
+def vs_installation_path(version, preference=None):
+
+    if not preference:
+        preference = [var.lstrip().rstrip() for var in get_env("CONAN_VS_INSTALLATION_PREFERENCE",
+                                                               list())]
+    
+    if not preference:
+        preference = ["Enterprise", "Professional", "Community", "BuildTools"]
+
     if not hasattr(vs_installation_path, "_cached"):
         vs_installation_path._cached = dict()
 
-    if version not in vs_installation_path._cached:
-        vs_path = None
-        legacy_products = vswhere(legacy=True)
+    if version in vs_installation_path._cached:
+        if vs_installation_path._cached[version] is not None:
+            if preference[0] in vs_installation_path._cached[version]:
```

Comment:
I am not sure I understand the logic. So if the cached one is not the first preference, we invalidate it? How it is possible that this would be different from previous run (env-vars should be the same). Maybe the ``_cached`` should cache by version AND preference?